src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -129,32 +93,48 @@ private void RemoveTagsThatIntersectEdit(TextContentChangedEventArgs e)
             if (e.Changes.Count == 0)
                 return;
 
-            var buffer = e.After.TextBuffer;
-            if (!this.CachedTagTrees.TryGetValue(buffer, out var treeForBuffer))
-                return;
-
             var snapshot = e.After;
+            var buffer = snapshot.TextBuffer;
 
-            var tagsToRemove = e.Changes.SelectMany(c => treeForBuffer.GetIntersectingSpans(new SnapshotSpan(snapshot, c.NewSpan)));
-            if (!tagsToRemove.Any())
-                return;
+            // Everything we're passing in here is synchronous.  So we can assert that this must complete synchronously
+            // as well.
+            var (oldTagTrees, newTagTrees) = CompareAndSwapTagTreesAsync(
```

Comment:
core helper is this CompareAndSwapTagTreesAsync function.  It performs the normal compare-and-swap algorithm, reading teh current value of _cachedTagTrees, calling a user-provided function with it, and attempting to then update that field with the new result as long as it didn't change in between.  If it did change, it spins until it finally wins and then returns the before/after state.